README.md
jQuery tableFilter
==================
*Requires [jQuery](http://jquery.com/)*  
*Requires [utils](https://github.com/FagnerMartinsBrack/jQuery-Table-Filter/blob/master/src/utils.js)*

* Given a table with thead and tbody, jQuery Table Filter creates an input below all TDs of theads that automatically filters the content based on the innerHTML of each TD of the tbody
* The TRs that don't match with desired filter are hidden and any input inside any TD is disabled (the input is enabled when the TR is visible again)
* The filter verifies the content splitting each word as it is a different search ("A test" matches with "A - test" and "test A")
* The only requirement for the use of this plugin is that it is used ONLY in a table with at least a &lt;thead&gt;, &lt;th&gt; and &lt;tbody&gt;
* It is created a TR that informs the user when a filter did not return any result. It is positioned below the last TR of the tfoot. If there's no tfoot, jQuery Table Filter creates it.

Parameters
----------
Parameters in "[]" are optional:
* {String} [inputClasses] - custom classes that will be set in the filter inputs that will be created
* {String} [trClasses] - custom classes that will be set in the TR(s) that will be created to contain the filters
* {String} [thClasses] - custom classes that will be set in the TH(s) that will be created to contain the filters
* {Number} [timeout=1000] - delay to apply the filter after typing (usefull if the table has a huge number of content)
* {Boolean} [ignoreCase=false] - Verifies if jQuery Table Filter should ignore the case
* {String} [emptyMessage="There's no result to show"] - Message when there's no result (A tfoot will be created if there's none)
* {Array} [columns] - Index of the columns that will be filtered. If not specified or it is an empty array consider all columns.

Usage
-----
<pre>

//Regular usage
$("#table").tableFilter();

//Passing parameters
$("#table").tableFilter({
    timeout: 0,
    columns: [0, 2, 3],
    emptyMessage: "Cannot find any occurences..."
});
</pre>

Changelog
--------
* 1.0.2 - Initial creation of README. Conform code with jQuery code style. Change some internal data/bind namespaces.

changelog.txt
1.1.0 - Allows the possibility to add a select input instead of the default text input

src/utils.js
Array.prototype.contains = function(obj) {
	var _reg = 0, t = this, i = t.length;
	while(i--) { //!== 0
		if(t[i] === obj) {
			_reg++;
		}
	}
return _reg;
};
String.prototype.replaceAll = function(token, newToken, ignoreCase) {
	var str, i = -1, _token;
	if((str = this.toString()) && typeof token === "string") {
		_token = ignoreCase === true? token.toLowerCase() : undefined;
		while((i = (
			_token !== undefined? 
				str.toLowerCase().indexOf(
							_token, 
							i >= 0? i + newToken.length : 0
				) : str.indexOf(
							token,
							i >= 0? i + newToken.length : 0
				)
		)) !== -1 ) {
		    str = str.substring(0, i)
		    		.concat(newToken)
		    		.concat(str.substring(i + token.length));
		}
	}
return str;
};
String.prototype.contains = function(token, ignoreCase) {
	var _reg = 0, str = this.toString(), i;
	if(str && typeof token === "string") {
		if(ignoreCase === true) {
			token = token.toLowerCase();
			str = str.toLowerCase();
		}
		while((i = str.indexOf(token)) !== -1) {
			str = str.substring(i + token.length);
			_reg++;
		}
	}
	return _reg;
};

src/JSTableFilter.js
/**
 * JS Table Filter 1.0 (jQuery 1.7+)
 * (c) 2012-2012 Fagner Martins Brack <fagnerbrack.com>
 * MIT license
 * 
 * JS Table Filter
 * Given a table with thead and tbody, JS Table Filter creates an input below all TDs of theads that automatically filters the content based on the innerHTML of each TD of the tbody
 * The TRs that don't match with desired filter are hidden and any input inside any TD is disabled (the input is enabled when the TR is visible again)
 * The filter verifies the content splitting each word as it is a different search ("A test" matches with "A - test" and "test A")
 * The only requirement for the use of this plugin is that it is used ONLY in a table with at least a <thead>, <th> and <tbody>
 * It is created a TR that informs the user when a filter did not return any result. It is positioned below the last TR of the tfoot. If there's no tfoot, JS Table Filter creates it.
 * 
 * Use:
 * $('table').syoTableFilter([Object parameters]);
 */
(function($) {
	$.fn.JSTableFilter = function(args) {
		var options = $.extend({
			inputClasses: undefined, //custom classes that will be set in the filter inputs that will be created
			trClasses: undefined, //custom classes that will be set in the TR(s) that will be created to the filters
			delay: 1000, //delay to apply the filter after typing (if the table has a huge number of content)
			ignoreCase: true, //Verifies if JS Table Filter should ignore the case
			emptyMessage: "There's no result to show", //Message when there's no result
			columns: [] //Index of the columns that will be filtered. If not especified consider all columns.
		}, args);
		
		function option(key) {
			var ret = undefined;
			if(key !== undefined) {
				ret = options[key];
			}
		return ret;
		};
		
		/**
		 * @param colNumber The number of TDs in the table's thead
		 * @return A jQuery reference to the new TR that will have the filters
		 */
		function getFilterTR(colNumber) {
			var inputClasses = option('inputClasses');
			var trClasses = option('trClasses');
			var columns = option('columns');
			var str = [];
			str.push("<tr class='" + ($.type(trClasses) === 'string'? trClasses : '') + "'>");
				for(var i = 0; i<colNumber; i++) {
					str.push("<td style='text-align: center; vertical-align: middle; padding: 0; margin: 0;'>");
						if(!$.isArray(columns) || !columns.length || columns.contains(i)) {
							str.push("<input type='text' class='" + ($.type(inputClasses) === 'string'? inputClasses : '') + "'/>");
						}
					str.push("</td>");
				}
			str.push("</tr>");
		return $(str.join(''));
		};
		
		/**
		 * @param colNumber Greater number of TDs in the table (to set the colspan)
		 * @return Um objeto jQuery reference to the TR that will appear in the footer when a filter shows no TR
		 */
		function getFooterTR(colNumber) {
			var str = [];
			str.push("<tr class='syo-filter-info'>");
				str.push("<td colspan='" + colNumber + "'>");
					str.push(options.emptyMessage);
				str.push("</td>");
			str.push("</tr>");
		return $(str.join(''));
		};
		
		/**
		 * Set the .data() in the TDs the value if its index
		 * So you dont need to verify the index via selector
		 */
		function markTDs() {
			var $this = $(this);
			$this.data('filter-index', $this.index());
		};
		
		function acceptKey(e) {
			var accept = true;
			if(e.keyCode in {
				16: 'shift',
				17: 'control',
				18: 'alt',
				32: 'space'
			}) {
				accept = false;
			}
		return accept;
		};
		
		/**
		 * @param $customTR jQuery reference to the new TR that will have the filters
		 * @param $bodyTRs jQuery reference to the table
		 */
		function bind($customTR, $table) {
			var $bodyTRs = $table.find('tbody tr');
			var $bodyTDs = $bodyTRs.find('td').each(markTDs);
			var $headTDs = $customTR.find('td');
			var $allInputs = $headTDs.find('input');
			var colNumber = $headTDs.length;
			$headTDs.each(function(index, element) {
				var $TD = $(this);
				
				//Get the input of this TD
				$TD.find('input').on('keydown', function(e) {
					if(acceptKey(e)) { //Verifies if the typed key is acceptable in the filter
						var $input = $(this);
						var ignoreCase = option('ignoreCase');
						var filterTimeout = $input.data('syo_filter_timeout');
						
						if(filterTimeout !== undefined) {
							clearTimeout(filterTimeout);
						}
						
						$input.data('syo_filter_timeout', setTimeout(function() {
							var $TRsToHide = $();
							show($bodyTRs);
							
							//Iterate between all inputs and filters each respective column
							$allInputs.each(function(_i, el) {
								var $input = $(el);
								var inputVal = $input.val().trim();
								var index;
								
								if(inputVal) {
									index = $input.parent().index(); //Get the index of the TD not the input (cause it is possible that not all TDs have inputs)
									$bodyTDs.each(function() {
										var $thisTD = $(this);
										var _index = $thisTD.data('filter-index');
										if(_index === index) {
											var innerText = $thisTD.text().trim();
											
											//If it did'nt match...
											if(!compare(innerText, inputVal, ignoreCase)) {
												$TRsToHide = $TRsToHide.add(this.parentNode); //...Add to hide the TR of this TD
											}
										}
									});
								}
							});
							
							//Hide every marked TR
							hide($TRsToHide);
							
							//Verifies if there's no TR to show
							var $tfoot = $table.find('tfoot:last');
							if(!$tfoot.length) $tfoot = $('<tfoot />').appendTo($table.find('tbody:last').parent());
							
							var $infoTR = $tfoot.find('tr.syo-filter-info');
							if($bodyTRs.filter(':visible').length === 0) {
								if(!$infoTR.length) {
									$tfoot.append(getFooterTR(colNumber)).show();
								} else {
									$infoTR.show();
								}
							} else {
								$infoTR.hide();
							}
						}, option('delay')));
					}
				});
			});
			return $customTR;
		};
		
		function hide($TRs) {
			$TRs.not(':hidden').each(function() {
				//Add a flag to the TRs that will be hidden and hide it
				var $this = $(this);
				$this.data('filter_hide', true);
				$this.hide();
			});
		};
		
		function show($TRs) {
			$TRs.each(function() {
				var $this = $(this);
				
				//Remove the flag of the hidden TRs and show it
				if($this.data('filter_hide') === true) {
					$this.removeData('filter_hide');
					$this.show();
				}
			});
		};
		
		function splitQuotes(str) {
			var reg = new RegExp('([ ]?")([^"*]*)("[ ]?)', 'g');
			var ret = [];
			var matches = null;
			while((matches = reg.exec(str)) !== null) {
				ret.push({
					0: matches[0],
					1: matches[2]
				});
			}
		return ret;
		};
		
		function splitValues(values) {
			
			//Remove double quote
			var quotes = splitQuotes(values);
			for(var i = 0; i<quotes.length; i++) {
				values = values.replaceAll(quotes[i][0], ' ');
			}
			
			//Prepare the return
			var trimmedVal = values.trim();
			var ret = trimmedVal? trimmedVal.split(' ') : [];
			for(var i = 0; i<quotes.length; i++) {
				ret.push(quotes[i][0].trim());
			}
			
			return ret;
		};
		
		/**
		 * Comparator method to check if the result of the filter matches with the column text
		 * @param {String} text The innerHTML of the TD
		 * @param {String} val The typed value to filter this column
		 * @param {Boolean} ignoreCase If the compare should be case-sensitive
		 */
		function compare(text, values, ignoreCase) {
			var valArray = splitValues(values);
			var c = 0;
			for(var i = 0; i<valArray.length; i++) {
				var val = valArray[i];
				var l = val.length;
				if(l >= 2 && val.substring(0, 1) === '"' && val.substring(l - 1, l) === '"') {
					val = val.substring(1, l - 1);
					if(contains(text, val, ignoreCase)) {
						c++;
					}
				} else if(text.contains(val, ignoreCase)) {
					c++;
				}
			}
			return c === valArray.length;
		};
		
		/**
		 * Check if the value matches with any word inside the string
		 */
		function contains(text, val, ignoreCase) {
			
			function hasMatch(index) {
				var matchLeft = index - 1 === -1 || text.substring(index - 1, index) === ' ';
				var matchRight = index + val.length >= text.length || text.substring(index + val.length, index + val.length + 1) === ' ';
				return matchLeft && matchRight;
			};
			
			return ignoreCase === true? hasMatch(text.toUpperCase().indexOf(val.toUpperCase())) : hasMatch(text.indexOf(val));
		};
		
		$(this).each(function() {
			var $table = $(this);
			var $thead = $table.find('thead:first');
			var colNumber = $thead.find('tr:last th').length;
			var $customTR = bind(getFilterTR(colNumber), $table);
			
			$thead.append($customTR);
		});
	return this;
	};
})(jQuery);

src/syoTableFilter.js
/**
 * JS Table Filter 1.0 (jQuery 1.7+)
 * (c) 2012-2012 Fagner Martins Brack <fagnerbrack.com>
 * MIT license
 * 
 * JS Table Filter
 * Given a table with thead and tbody, JS Table Filter creates an input below all TDs of theads that automatically filters the content based on the innerHTML of each TD of the tbody
 * The TRs that don't match with desired filter are hidden and any input inside any TD is disabled (the input is enabled when the TR is visible again)
 * The filter verifies the content splitting each word as it is a different search ("A test" matches with "A - test" and "test A")
 * The only requirement for the use of this plugin is that it is used ONLY in a table with at least a <thead>, <th> and <tbody>
 * It is created a TR that nforms the user when a filter did not return any result. It is positioned below the last TR of the tfoot. If there's no tfoot, JS Table FIlter creates it.
 * 
 * Use:
 * $('table' ).syoTableFilter([Object parameters]);
 */
(function($) {
	$.fn.syoTableFilter = function(args) {
		var options = $.extend({
			inputClasses: undefined, //custom classes that will be set in the filter inputs that will be created
			trClasses: undefined, //custom classes that will be set in the TR(s) that will be created to the filters
			delay: 1000, //delay to apply the filter after typing (if the table has a huge number of content)
			ignoreCase: true, //Verifies if JS Table Filter should ignore the case
			emptyMessage: "There's no result to show", //Message when there's no result
			columns: [] //Index of the columns that will be filtered. If not especified consider all columns.
		}, args);
		
		function option(key) {
			var ret = undefined;
			if(key !== undefined) {
				ret = options[key];
			}
		return ret;
		};
		
		/**
		 * @param colNumber The number of TDs in the table's thead
		 * @return A jQuery reference to the new TR that will have the filters
		 */
		function getFilterTR(colNumber) {
			var inputClasses = option('inputClasses');
			var trClasses = option('trClasses');
			var columns = option('columns');
			var str = [];
			str.push("<tr class='" + ($.type(trClasses) === 'string'? trClasses : '') + "'>");
				for(var i = 0; i<colNumber; i++) {
					str.push("<td style='text-align: center; vertical-align: middle; padding: 0; margin: 0;'>");
						if(!$.isArray(columns) || !columns.length || columns.contains(i)) {
							str.push("<input type='text' class='" + ($.type(inputClasses) === 'string'? inputClasses : '') + "'/>");
						}
					str.push("</td>");
				}
			str.push("</tr>");
		return $(str.join(''));
		};
		
		/**
		 * @param colNumber Greater number of TDs in the table (to set the colspan)
		 * @return Um objeto jQuery reference to the TR that will appear in the footer when a filter shows no TR
		 */
		function getFooterTR(colNumber) {
			var str = [];
			str.push("<tr class='syo-filter-info'>");
				str.push("<td colspan='" + colNumber + "'>");
					str.push(options.emptyMessage);
				str.push("</td>");
			str.push("</tr>");
		return $(str.join(''));
		};
		
		/**
		 * Set the .data() in the TDs the value if its index
		 * So you dont need to verify the index via selector
		 */
		function markTDs() {
			var $this = $(this);
			$this.data('filter-index', $this.index());
		};
		
		function acceptKey(e) {
			var accept = true;
			if(e.keyCode in {
				16: 'shift',
				17: 'control',
				18: 'alt',
				32: 'space'
			}) {
				accept = false;
			}
		return accept;
		};
		
		/**
		 * @param $customTR jQuery reference to the new TR that will have the filters
		 * @param $bodyTRs jQuery reference to the table
		 */
		function bind($customTR, $table) {
			var $bodyTRs = $table.find('tbody tr');
			var $bodyTDs = $bodyTRs.find('td').each(markTDs);
			var $headTDs = $customTR.find('td');
			var $allInputs = $headTDs.find('input');
			var colNumber = $headTDs.length;
			$headTDs.each(function(index, element) {
				var $TD = $(this);
				
				//Get the input of this TD
				$TD.find('input').on('keydown', function(e) {
					if(acceptKey(e)) { //Verifies if the typed key is acceptable in the filter
						var $input = $(this);
						var ignoreCase = option('ignoreCase');
						var filterTimeout = $input.data('syo_filter_timeout');
						
						if(filterTimeout !== undefined) {
							clearTimeout(filterTimeout);
						}
						
						$input.data('syo_filter_timeout', setTimeout(function() {
							var $TRsToHide = $();
							show($bodyTRs);
							
							//Iterate between all inputs and filters each respective column
							$allInputs.each(function(_i, el) {
								var $input = $(el);
								var inputVal = $input.val().trim();
								var index;
								
								if(inputVal) {
									index = $input.parent().index(); //Get the index of the TD not the input (cause it is possible that not all TDs have inputs)
									$bodyTDs.each(function() {
										var $thisTD = $(this);
										var _index = $thisTD.data('filter-index');
										if(_index === index) {
											var innerText = $thisTD.text().trim();
											
											//If it did'nt match...
											if(!compare(innerText, inputVal, ignoreCase)) {
												$TRsToHide = $TRsToHide.add(this.parentNode); //...Add to hide the TR of this TD
											}
										}
									});
								}
							});
							
							//Hide every marked TR
							hide($TRsToHide);
							
							//Verifies if there's no TR to show
							var $tfoot = $table.find('tfoot:last');
							if(!$tfoot.length) $tfoot = $('<tfoot />').appendTo($table.find('tbody:last').parent());
							
							var $infoTR = $tfoot.find('tr.syo-filter-info');
							if($bodyTRs.filter(':visible').length === 0) {
								if(!$infoTR.length) {
									$tfoot.append(getFooterTR(colNumber)).show();
								} else {
									$infoTR.show();
								}
							} else {
								$infoTR.hide();
							}
						}, option('delay')));
					}
				});
			});
			return $customTR;
		};
		
		function hide($TRs) {
			$TRs.not(':hidden').each(function() {
				//Add a flag to the TRs that will be hidden and hide it
				var $this = $(this);
				$this.data('filter_hide', true);
				$this.hide();
			});
		};
		
		function show($TRs) {
			$TRs.each(function() {
				var $this = $(this);
				
				//Remove the flag of the hidden TRs and show it
				if($this.data('filter_hide') === true) {
					$this.removeData('filter_hide');
					$this.show();
				}
			});
		};
		
		function splitQuotes(str) {
			var reg = new RegExp('([ ]?")([^"*]*)("[ ]?)', 'g');
			var ret = [];
			var matches = null;
			while((matches = reg.exec(str)) !== null) {
				ret.push({
					0: matches[0],
					1: matches[2]
				});
			}
		return ret;
		};
		
		function splitValues(values) {
			
			//Remove double quote
			var quotes = splitQuotes(values);
			for(var i = 0; i<quotes.length; i++) {
				values = values.replaceAll(quotes[i][0], ' ');
			}
			
			//Prepare the return
			var trimmedVal = values.trim();
			var ret = trimmedVal? trimmedVal.split(' ') : [];
			for(var i = 0; i<quotes.length; i++) {
				ret.push(quotes[i][0].trim());
			}
			
			return ret;
		};
		
		/**
		 * Comparator method to check if the result of the filter matches with the column text
		 * @param {String} text The innerHTML of the TD
		 * @param {String} val The typed value to filter this column
		 * @param {Boolean} ignoreCase If the compare should be case-sensitive
		 */
		function compare(text, values, ignoreCase) {
			var valArray = splitValues(values);
			var c = 0;
			for(var i = 0; i<valArray.length; i++) {
				var val = valArray[i];
				var l = val.length;
				if(l >= 2 && val.substring(0, 1) === '"' && val.substring(l - 1, l) === '"') {
					val = val.substring(1, l - 1);
					if(contains(text, val, ignoreCase)) {
						c++;
					}
				} else if(text.contains(val, ignoreCase)) {
					c++;
				}
			}
			return c === valArray.length;
		};
		
		/**
		 * Check if the value matches with any word inside the string
		 */
		function contains(text, val, ignoreCase) {
			
			function hasMatch(index) {
				var matchLeft = index - 1 === -1 || text.substring(index - 1, index) === ' ';
				var matchRight = index + val.length >= text.length || text.substring(index + val.length, index + val.length + 1) === ' ';
				return matchLeft && matchRight;
			};
			
			return ignoreCase === true? hasMatch(text.toUpperCase().indexOf(val.toUpperCase())) : hasMatch(text.indexOf(val));
		};
		
		$(this).each(function() {
			var $table = $(this);
			var $thead = $table.find('thead:first');
			var colNumber = $thead.find('tr:last th').length;
			var $customTR = bind(getFilterTR(colNumber), $table);
			
			$thead.append($customTR);
		});
	return this;
	};
})(jQuery);

src/jquery.tableFilter.js
/*!
 * jQuery Table Filter 1.1.0 (jQuery 1.7+)
 * Fagner Martins Brack <fagnerbrack.com>
 * MIT license
 * https://github.com/FagnerMartinsBrack/jQuery-Table-Filter
 */
(function( $, undefined ) {
	$.fn.tableFilter = function(args) {
		var options = $.extend({
			
			//delay to apply the filter after typing (if the table has a huge number of content)
			delay: 1000,
			
			//Verifies if jQuery Table Filter should ignore the case
			ignoreCase: false,
			
			//Message when there's no result
			emptyMessage: "There's no result to show"
			
		}, args);
		
		function option(key) {
			return options[key];
		}
		
		/**
		 * @param colNumber The number of TDs in the table's thead
		 * @returns {jQuery} A jQuery reference to the new TR that will have the filters
		 */
		function getFilterTR(colNumber) {
			var i = 0,
				trClasses = option('trClasses'),
				thClasses = option('thClasses'),
				columns = option('columns'),
				str = [];
			
			str.push("<tr class='" + (typeof trClasses === "string" ? trClasses : "") + "'>");
				for(; i<colNumber; i++) {
					str.push("<th class='" + (typeof thClasses === "string" ? thClasses : "") + "'>");
						if( !columns || $.isArray(columns) && columns.contains(i) ) {
							str.push( getInputFilter("text") );
						} else if( $.isPlainObject(columns) ) {
							str.push( getInputFilter(columns[i]) );
						}
					str.push("</th>");
				}
			str.push("</tr>");
		return $( str.join("") );
		}
		
		/**
		 * @param {String} inputType The type of input to be created
		 * @returns {String} A string representing the input structure to be created
		 */
		function getInputFilter(inputType) {
			
			var ret,
				inputClasses = option('inputClasses') || "";
			
			if(inputType === "text") {
				ret = "<input type='text' class='" + inputClasses + "'/>"
			} else if(inputType === "select-one") {
				ret = "<select type='select-one' class='" + inputClasses + "'></select>";
			}
			
		return ret;
		}
		
		/**
		 * @param colNumber Greater number of TDs in the table (to set the colspan)
		 * @returns {jQuery} A jQuery reference to the TR that will appear in the footer when a filter shows no TR
		 */
		function getFooterTR(colNumber) {
			var str = [];
			str.push("<tr class='filter-info'>");
				str.push("<td colspan='" + colNumber + "'>");
					str.push( option("emptyMessage") );
				str.push("</td>");
			str.push("</tr>");
		return $( str.join("") );
		}
		
		/**
		 * Store the TD index to avoid a DOM check every time
		 */
		function eachTDs() {
			var $td = $(this);
			$td.data( "tablefilter-index", $td.index() );
		}
		
		/**
		 * Given a jQuery event object, return false if the key pressed matches with the map
		 * 
		 * @param {Object} The Event Object
		 */
		function acceptKey(e) {
			return !(e.keyCode in {
				16: "shift",
				17: "control",
				18: "alt",
				32: "space"
			});
		}
		
		/**
		 * Bind the events to the inputs and create the select input options
		 * 
		 * @param $customTR jQuery reference to the new TR that will have the filters
		 * @param $bodyTRs jQuery reference to the table
		 * 
		 * @returns {jQuery} The $customTR as a jQuery Object
		 */
		function bind( $customTR, $table ) {
			var $bodyTRs = $table.find("tbody tr"),
				$bodyTDs = $bodyTRs.find("td").each(eachTDs),
				$headTHs = $customTR.find("th"),
				$allInputs = $headTHs.find("input, select"),
				colNumber = $headTHs.length,
				ignoreCase = option("ignoreCase"),
				delay = option("delay");
			
			$headTHs.each(function( index, element ) {
				
				var $TH = $(this);
				
				function _applyFilter() {
					var innerText,
						$tfoot, $infoTR,
						$TRsToHide = $();
					
					show($bodyTRs);
					
					//Iterate between all inputs and filters each respective column
					$allInputs.each(function(_i) {
						var index,
							$input = $(this),
							inputVal = $input.val().trim();
						
						if(inputVal) {
							
							//Get the index of the TD not the input (cause it is possible that not all TDs have inputs)
							index = $input.parent().index();
							
							$bodyTDs.each(function() {
								var $thisTD = $(this),
									_index = $thisTD.data("tablefilter-index");
								
								if( _index === index ) {
									innerText = $thisTD.text().trim();
									
									//If it did'nt match...
									if( !compare(innerText, inputVal, ignoreCase) ) {
										
										//...Add to hide the TR of this TD
										$TRsToHide = $TRsToHide.add( this.parentNode );
										
									}
								}
							});
						}
					});
					
					//Hide every marked TR
					hide($TRsToHide);
					
					//Verifies if there's no TR to show
					$tfoot = $table.find("tfoot:last");
					if( !$tfoot.length ) {
						$tfoot = $("<tfoot />").appendTo( $table.find('tbody:last').parent() );
					}
					
					$infoTR = $tfoot.find("tr.filter-info");
					if( !$bodyTRs.filter(":visible").length ) {
						if( !$infoTR.length ) {
							$tfoot.append( getFooterTR(colNumber) ).show();
						} else {
							$infoTR.show();
						}
					} else {
						$infoTR.hide();
					}
				}
				
				function bindText(e) {
					var $input;
					
					if( acceptKey(e) ) { //Verifies if the typed key is acceptable in the filter
						$input = $(this);
						initFilter($input)
					}
				}
				
				function bindSelect() {
					initFilter($(this));
				}
				
				function initFilter($input) {
					var filterTimeout = $input.data("tablefilter-timeout");
					
					if( filterTimeout !== undefined ) {
						clearTimeout(filterTimeout);
					}
					
					//Set the timeout only if the delay is greater than zero
					if( delay > 0 ) {
						$input.data( "tablefilter-timeout", setTimeout(_applyFilter, delay) );
					} else {
						_applyFilter();
					}
				}
				
				function createOptions($selectElement) {
						var i = 0,
							arrOptions = [],
							options = "";
					
					if($selectElement.length) {
						$bodyTDs.each(function() {
							var innerText,
								$thisTD = $(this),
								_index = $thisTD.data("tablefilter-index");
							
							if( _index === index ) {
								innerText = $thisTD.text().trim();
								
								if( $.inArray(innerText, arrOptions) === -1 ) {
									arrOptions.push(innerText);
								}
							}
						});
						
						//alphabetical
						arrOptions.sort(function(a, b) {
							return a > b;
						});
						
						options += "<option selected='selected'></option>";
						
						for( ; i<arrOptions.length; i++ ) {
							options += "<option value='" + arrOptions[i] + "'>" + arrOptions[i] + "</option>";
						}
						
						$selectElement.append(options);
					}
					return $selectElement;
				}
				
				//input
				//Change event to keyup. Sometimes a backspace keydown caused the input to be cleared after the code has been executed
				$TH.find("input").bind( "keyup.tablefilter", bindText);
				
				//select
				createOptions($TH.find("select")).bind( "change.tableFilter", bindSelect );
			});
		return $customTR.data( "tablefilter-tr", true );
		}
		
		/**
		 * Add a flag to the TRs that will be hidden and hide it
		 * 
		 * @param {jQuery} A jQuery Object with the TRs to be hidden
		 */
		function hide($TRs) {
			$TRs.not(":hidden").each(function() {
				var $tr = $(this)
					.data( "tablefilter-hide", true );
				
				$tr.find("input, select, textarea").each(function() {
					var $input = $(this);
					if( !$input.prop("disabled") ) {
						$input.data( "tablefilter-disabled", true ).prop( "disabled", true );
					}
				});
				
			$tr.hide();
			});
		}
		
		/**
		 * Show the given TRs
		 * 
		 * @param {jQuery} A jQuery Object with the TRs to be hidden
		 */
		function show($TRs) {
			$TRs.each(function() {
				var $tr = $(this);
				
				//Remove the flag of the hidden and disabled elements
				$tr.find("input, select, textarea").each(function() {
					var $input = $(this);
					if($input.data("tablefilter-disabled")) {
						$input.prop( "disabled", false ); //Just to clarify we should not use removeProp on native properties such as disabled
					}
				});
				
				//Remove the flag of the hidden TRs and show it
				if($tr.data("tablefilter-hide")) {
					$tr.removeData("tablefilter-hide");
					$tr.show();
				}
			});
		}
		
		/**
		 * Separate the quoted search from the general search
		 * 
		 * @param {String} str The string to be tested
		 * @returns {Array}
		 */
		function splitQuotes(str) {
			var matches,
				reg = new RegExp("([ ]?\")([^\"*]*)(\"[ ]?)", "g"),
				ret = [];
			
			while( (matches = reg.exec(str)) !== null ) {
				ret.push({
					0: matches[0],
					1: matches[2]
				});
			}
		return ret;
		}
		
		/**
		 * Split to treat each search as a separate one
		 * 
		 * @param {String} values The content of the search
		 * @returns {String} An array with each independent search
		 */
		function splitValues(values) {
			
			var i, trimmedVal,
				ret = [],
				quotes = splitQuotes(values);
			
			//Remove double quote
			for(i = 0; i<quotes.length; i++) {
				values = values.replaceAll( quotes[i][0], " " );
			}
			
			//Prepare the return
			if( trimmedVal = values.trim() ) {
				ret = trimmedVal.split(" ");
			}
			
			for(i = 0; i<quotes.length; i++) {
				ret.push( quotes[i][0].trim() );
			}
			
		return ret;
		}
		
		/**
		 * Comparator method to check if the result of the filter matches with the column text
		 * 
		 * @param {String} text The innerHTML of the TD
		 * @param {String} val The typed value to filter this column
		 * @param {Boolean} ignoreCase If the compare should be case-sensitive
		 * @returns {Boolean} If the values match with the compare rule within the given text
		 */
		function compare( text, values, ignoreCase ) {
			var val, l,
				c = 0,
				i = 0,
				valArray = splitValues(values);
			
			for(; i<valArray.length; i++) {
				
				val = valArray[i];
				l = val.length;
				
				if ( l >= 2 && val.substring(0, 1) === "\"" && val.substring(l - 1, l) === "\"" ) {
					
					val = val.substring(1, l - 1);
					if( contains(text, val, ignoreCase) ) {
						c++;
					}
					
				} else if( text.contains(val, ignoreCase) ) {
					c++;
				}
			}
		return c === valArray.length;
		}
		
		/**
		 * Check if the value matches with any word inside the string
		 * 
		 * @param {String} text The content of the TD
		 * @param {String} val The value to be compared
		 * @param {Boolean} ignoreCase if the compare should be made ignoring the case
		 * @returns {Boolean} If the given val is contained inside the text
		 */
		function contains( text, val, ignoreCase ) {
			
			function _hasMatch(index) {
				var matchLeft = index - 1 === -1 || text.substring( index - 1, index ) === " ",
					matchRight = index + val.length >= text.length || text.substring( index + val.length, index + val.length + 1 ) === " ";
				
				return matchLeft && matchRight;
			};
			
			return ignoreCase ? _hasMatch( text.toUpperCase().indexOf(val.toUpperCase()) ) : _hasMatch( text.indexOf(val) );
		};
		
		$(this).each(function() {
			var $table = $(this),
				$thead = $table.find("thead:first"),
				colNumber = $thead.find("tr:last th").length,
				$customTR = bind( getFilterTR(colNumber), $table ),
				$lastHeadTR = $thead.find("tr:last");
			
			//If a tr created by another instance of this plugin already exists, remove it.
			if( $lastHeadTR.data("tablefilter-tr") ) {
				$lastHeadTR.remove();
			}
			
			//Append the generated TR to the head
			$thead.append($customTR);
		});
	return this;
	};
})(jQuery);